### ISSUE TITLE
RL: DQN gives all 20 agents the same reward target (credit assignment)

### ISSUE THREAD

**Issue by VadimSokolov** [VadimSokolov/sandtable#6]
Nice work getting a full RL loop running against the simulator. This is about one specific thing in the DQN trainer that will cap what it can learn no matter how long it trains, and it is worth fixing before spending real compute.

## The problem

`src/sandtable/rl/train_dqn/train.py` has this target:

```python
b_reward = torch.FloatTensor(b_reward).unsqueeze(1)   # (batch, 1)
...
max_next_q = next_q.max(dim=2)[0]                     # (batch, n_agents)
# We distribute the global reward to all agents equally
target_q = b_reward + gamma * max_next_q * (1 - b_done)
loss = loss_fn(current_q_taken, target_q)
```

The `(batch, 1)` reward broadcasts across the agent axis, so all 20 agents are regressed onto the same scalar return. That means every head `Q_i(s, a_i)` is being trained to predict the whole team's return, and the target carries no information about whether agent `i`'s action was the one that helped.

With `n_agents = 20` and a reward that is mostly a single terminal +1000, agent `i`'s actual contribution is roughly 1/20 of the signal and the other 19 agents' choices are noise on top of it. The gradient that distinguishes "agent 3 went left" from "agent 3 went right" has to fight through the variance of 19 other decisions. This is the standard multi-agent credit assignment problem, and it does not go away with more episodes; it gets worse as `n_agents` grows.

There is a related consistency issue. Taking an independent `argmax` per agent (`q_vals.argmax(dim=2)` in the action-selection path) is only guaranteed to recover the best joint action if the joint value decomposes in a way the per-agent maxima respect. Regressing each head onto the full return does not give you that property, so the greedy joint action the trainer picks is not necessarily the greedy action of the value function it learned.

## Options, cheapest first

**1. VDN (value decomposition network).** Define the joint value as the sum of the per-agent values and regress *that* onto the reward:

```python
q_total      = current_q_taken.sum(dim=1, keepdim=True)          # (batch, 1)
next_q_total = max_next_q.sum(dim=1, keepdim=True)               # (batch, 1)
target       = b_reward + gamma * next_q_total * (1 - b_done)
loss         = loss_fn(q_total, target.detach())
```

This is a few lines and it buys the consistency property: under additivity, the per-agent argmax *is* the joint argmax, so the action selection you already have becomes correct. Worth trying first because it is nearly free. Sunehag et al. 2017, arXiv:1706.05296.

**2. QMIX.** Same idea, but replace the sum with a mixing network that is monotone in the per-agent values and conditioned on global state. Strictly more expressive than VDN and still keeps the argmax property. Rashid et al. 2018, arXiv:1803.11485.

**3. Difference rewards / COMA.** Give agent `i` the credit `R(a) - R(a_{-i}, default_i)`: what the team got, minus what it would have got had `i` done something default. Usually too expensive to compute, but SandTable runs a mission in about 0.2 s, so counterfactual rollouts are actually affordable here. This is arguably the most interesting option for us given the simulator we have. Foerster et al. 2017, arXiv:1705.08926.

## Two smaller things in the same file

**Reward sparsity.** The signal is `-0.1` per step and `+1000` on reaching the goal. Until the swarm reaches the goal by chance at least once, the only gradient is the step penalty, which teaches it to end episodes rather than to succeed. Potential-based shaping on distance-to-goal would give a dense signal without changing the optimal policy (Ng, Harada and Russell 1999). It composes with any of the three options above.

**Reward scale.** `+1000` against `-0.1` per step is a ratio of 10^4. Combined with `gamma` and long episodes, the terminal reward may not propagate back far enough to shape early movement. Worth normalizing the scales, or checking empirically how far back the value function actually carries.

## Suggested order

1. Add potential-based shaping so there is a learning signal before the first success.
2. Switch the loss to VDN. Small diff, and it fixes the argmax consistency issue at the same time.
3. Measure. If coordination is still the bottleneck, go to QMIX.
4. Keep difference rewards in mind as the option this simulator is unusually well suited for.

Happy to pair on any of these.
